WinUIGallery/ControlPages/ScratchPadPage.xaml.cs

```diff
@@ -0,0 +1,575 @@
+                    {
+                        var tagStartIndex = newText.LastIndexOf('<', selectionIndex - 1);
+                        // If we found a start tag and there wasn't already a close of that tag
+                        if (tagStartIndex >= 0 && newText.LastIndexOf('>', selectionIndex - 2) < tagStartIndex)
+                        {
+                            var tagName = newText.Substring(tagStartIndex + 1);
+
+                            char[] chars = { ' ', '/', '>', '\t', '\r', '\n' };
+                            var nameEndIndex = tagName.IndexOfAny(chars);
+                            if (nameEndIndex > 0)
+                            {
+                                tagName = tagName = tagName.Substring(0, nameEndIndex);
+                                if (tagName != "!--") // don't add a close tag for a comment
+                                {
+                                    InsertTextboxText("</" + tagName + ">", false);
+                                }
+                            }
+                        }
+                    }
+                    else if (newText[selectionIndex - 1] == '=' &&
+                        (selectionIndex >= newText.Length - 1 || newText[selectionIndex] != '"'))
+                    {
+                        // Might want to auto-insert quotes for a property. Check if this appears
+                        // to be inside a tag and just after a property name.
+                        char[] tagChars = { '<', '>' };
+                        var lastTagIndex = newText.LastIndexOfAny(tagChars, selectionIndex);
+                        if (lastTagIndex >= 0 && newText[lastTagIndex] == '<')
+                        {
+                            // In a tag. Make sure we aren't in a property value (improper comparison
+                            // here is to check if the last quote is proceed by an '=').
+                            var quoteIndex = newText.LastIndexOf('"', selectionIndex);
+                            if (quoteIndex < lastTagIndex || newText[quoteIndex - 1] != '=')
+                            {
+                                InsertTextboxText("\"\"", false);
+                                textbox.TextDocument.Selection.StartPosition++;
+                            }
+                        }
+
+                    }
+                }
+            }
+
+            // Save the text so next time we can compare against the new text
+            textbox.TextDocument.GetText(Microsoft.UI.Text.TextGetOptions.None, out m_oldText);
+        }
+
+        private void textbox_ActualThemeChanged(FrameworkElement sender, object args)
+        {
+            // Updating the formating for theme change
+            var formatter = new XamlTextFormatter(textbox);
+            formatter.ApplyColors();
+        }
+    }
+
+    public class XamlTextFormatter

```

Comment:
Did you consider ICSharpCode's controls?  they have rich support for syntax highlighting, indent, etc.

Reply:
I didn't want to add any any dependencies just for Scratch Pad. I hadn't intended to add any editor functionality, but ended up deciding that some quality-of-life improvements were needed.

For ICSharpCode, I see a WinForms and maybe WPF version, but I don't see any indicator of a drop-in solution for WinUI3.

ColorCode.WinUI, which Gallery already depends on, could be a future option. Currently it only works with RichTextBlock, which isn't very useful for an edit box scenario. But it may be extendable to RichEditBox. Likely RichEditBox would need to have some added features to address the undo stack issue mentioned in the PR description.